### DISCUSSION TITLE
💾💧 Memory leak

### DISCUSSION THREAD

**Discussion by moshie** [vercel/next.js#34188] [Help]
Hi all 👋🏻,

Strap in because this is a long one... 👨‍💻

## Introduction

To start I want to express I am looking to get some opinions & help from people more experienced with memory issues in NodeJS / NextJS.

I also want to mention I am not an expert at memory issues the following message expresses my findings and questions on the topic.

### Specifications

NextJS: v12.0.10
Local Node: v16.14.0
Docker Node: v16.13.2
PNPM: v6.28.0

## What's the issue?

We host a NextJS application on kubernetes we started noticing that the pods would crash and produce the following in the logs:

```
<--- Last few GCs --->
[261:0x7f3a2a4bf130] 99059389 ms: Mark-sweep (reduce) 2006.5 (2052.8) -> 2006.5 (2052.3) MB, 3771.9 / 6.5 ms  (average mu = 0.918, current mu = 0.000) last resort GC in old space requested
[261:0x7f3a2a4bf130] 99063314 ms: Mark-sweep (reduce) 2006.5 (2052.3) -> 2006.5 (2052.3) MB, 3925.0 / 8.4 ms  (average mu = 0.846, current mu = 0.000) last resort GC in old space requested
<--- JS stacktrace --->
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
```

Once the pods had restarted we monitored the memory and noticed that over the space of roughly a day each pod would progressively increase in memory from 200MB before crashing at the limit set by kubernetes (4GB).

### Production:

<img width="1649" alt="Screenshot 2022-02-10 at 15 35 28" src="https://user-images.githubusercontent.com/3974301/153441378-1f171af7-283a-4605-bc73-f2e77f618683.png">

<img width="704" alt="Screenshot 2022-02-10 at 15 55 00" src="https://user-images.githubusercontent.com/3974301/153445266-854b1160-2dcd-45ef-b144-0543cba7544d.png">

## Application Background

The application is very simple it sits behind authentication and its sole purpose is to show information cards to a user so that they can be navigated to other team's applications.

The cards are pulled down using an api request from the client to `/api/cards` from here it makes 4 requests out to other apis in order to get the card information from various teams this data is then concatenated together then sent back to the client.

The application also handles the ability to change users personal data email, name, phone number ect.

The site averages 20,000 - 28,000 users a day.

## What have you tried

I haven't necessarily used memory inspection tools in great detail before so this was new for me..

First thing we tried was to see if we could replicate the leak locally

At first glance (No traffic) dev tools is reporting the in use heap size at 33.1MB.
In the activity monitor one of the two node processes is reported to be at 274.2MB in memory the other at 66.4MB.

On first visit of the site (Homepage only) the memory on the mac activity monitor increases from 274.2MB to 288.6MB (+14.4MB) the in use heap size is 33.1MB

On the second visit of the node process in the activity monitor goes up again to 297.9 (+9.3MB) however the heap size stays the same at 33.1MB.

On third visit it increases to 312MB (+14.1MB) Heap still at 33.1MB.

I left some time between each visit to give the memory a chance to settle before taking note of the value.

The memory never goes down between changes it always increases.  📈 

When running `process.memoryUsage()` locally I get the following:

```js
{rss: 101584896, heapTotal: 35713024, heapUsed: 33706512, external: 1348446, arrayBuffers: 283369}
```

Why is the RSS so small compared to the process in activity monitor?

Because the heap stays the same size throughout the test I guess doing a heap snapshot won't yield much in the way of useful results? (I am not 100% sure on this so please tell me if I am wrong).

I did one anyways I can't share it all but let me know if you want me to drill down further into any of these:

<img width="953" alt="Screenshot 2022-02-10 at 16 08 22" src="https://user-images.githubusercontent.com/3974301/153447791-d179080e-4215-4c83-8d7f-a8f9864d6cff.png">

I did find [this article](https://github.com/vercel/next.js/discussions/26801#discussioncomment-1091592) but adding: `experimental: { isrMemoryCacheSize: 0 },` didn't help.

I then moved on from this and tried using kubectl to inspect the node process of our staging environment. I managed to get at the console and type a few commands:

```js
// process.memoryUsage();
{rss: 165564416, heapTotal: 71540736, heapUsed: 66266672, external: 3182269, arrayBuffers: 4619657}
{rss: 172294144, heapTotal: 73056256, heapUsed: 68522432, external: 1903979, arrayBuffers: 4849625} // After 5 mins
{rss: 172605440, heapTotal: 73056256, heapUsed: 69137200, external: 1903995, arrayBuffers: 4849641} // immediately after the previous one
{rss: 178196480, heapTotal: 72851456, heapUsed: 66142240, external: 3290508, arrayBuffers: 4475957} // After first view
```

```js
// v8.getHeapStatistics()
{
  does_zap_garbage: 0
  heap_size_limit: 1660944384
  malloced_memory: 1065032
  number_of_detached_contexts: 0
  number_of_native_contexts: 1
  peak_malloced_memory: 17259720
  total_available_size: 1593436032
  total_heap_size: 71540736
  total_heap_size_executable: 3145728
  total_physical_size: 70831656
  used_heap_size: 67235784
}
```

Staging Pods:

<img width="425" alt="Screenshot 2022-02-10 at 16 13 25" src="https://user-images.githubusercontent.com/3974301/153448935-273476f6-8a98-4f11-a637-dc7ca0736d52.png">

With each visit of the staging site the memory increases.

Been banging my head against a wall on this one for sure. 

Please let me know what you would like me to try to find the source of the leak. 🙏

**Comment by balazsorban44** [comment]
Hi, out of curiosity, are you using `next-transpile-modules`?

**Reply by moshie** [reply]
Hey, nope I am not using that package.

**Reply by weyert** [reply]
I am having similar issues with recent Next.js version (12.0.7), the memory usages also goes up. I do use `next-transpile-modules`. @balazsorban44 You might think it would be related?

I visited one page and then not visiting any other browsers.  Just taking snapshots via the attached debugger and the suggested memory usage is as follows:

```
97mb  15:24:35
138mb 16:08:20 difference: 41mb
163mb 16:36:10 difference: 25mb
187mb 17:01:38 difference: 24mb
217mb 17:33:01 difference: 30mb
235mb 17:53:55 difference: 18mb
254mb 18:17:21 difference: 19mb
```

**Comment by moshie** [comment]
I did some more digging this time did a timeline and noticed that nodeNextRequest is being held on to?

<img width="1425" alt="Screenshot 2022-02-14 at 14 38 43" src="https://user-images.githubusercontent.com/3974301/153884664-00758d26-dbbc-4920-b026-1f1aafdf6ed0.png">

<img width="1679" alt="Screenshot 2022-02-14 at 15 04 28" src="https://user-images.githubusercontent.com/3974301/153889159-de04214d-6357-4bd1-8e8e-076b5c2f219e.png">

**Reply by fmmoret** [reply]
are you using next-connect? Bug in versions <0.12.1 https://github.com/hoangvvo/next-connect/pull/178

**Reply by moshie** [reply]
Nope not using that

**Comment by moshie** [comment]
After so much time spent restarting pods this is finally solved!!

Updating to version `12.3.0` fixed the issue.

It seems that this PR solved it in the end https://github.com/oscard0m/web/pull/203
